vpr/src/place/place.cpp

```diff
@@ -670,30 +793,64 @@ void try_place(const t_placer_opts& placer_opts,
                                       pin_timing_invalidator.get(),
                                       timing_info.get());
 
-        placement_inner_loop(&state, placer_opts,
-                             inner_recompute_limit, &stats,
-                             &costs,
-                             &moves_since_cost_recompute,
-                             pin_timing_invalidator.get(),
-                             place_delay_model.get(),
-                             placer_criticalities.get(),
-                             placer_setup_slacks.get(),
-                             *move_generator,
-                             blocks_affected,
-                             timing_info.get(),
-                             placer_opts.place_algorithm);
-
-        tot_iter += state.move_lim;
-        ++state.num_temps;
-
         if (placer_opts.place_algorithm.is_timing_driven()) {
             critical_path = timing_info->least_slack_critical_path();
             sTNS = timing_info->setup_total_negative_slack();
             sWNS = timing_info->setup_worst_negative_slack();
+
+            //see if we should save the current placement solution as a checkpoint
+            if (placer_opts.place_checkpointing && agent_state == 2) {
+                if (placement_checkpoint.cp_is_valid() == false || (timing_info->least_slack_critical_path().delay() < placement_checkpoint.get_cp_cpd() && costs.bb_cost <= placement_checkpoint.get_cp_bb_cost())) {
+                    placement_checkpoint.save_placement(costs, critical_path.delay());
+                    VTR_LOG("Checkpoint saved: bb_costs=%g, TD costs=%g, CPD=%7.3f (ns) \n", costs.bb_cost, costs.timing_cost, 1e9 * critical_path.delay());
+                }
+            }
         }
 
+        if (agent_state == 1 || placer_opts.place_agent_multistate == false) {
```

Comment:
This is a lot of lines of code to just pass in a different *move_generator.  Could do with a shorter if statement to set a variable (
if (...)
   current_move_generator = move_generator;
else {
   current_move_generator = move_generator2;

or some such.

Reply:
Done by creating a new helper function. In this case, I had to create two helper functions as move_generator is a unique_ptr so I am moving it then moving it back after running placement_inner_loop.